src/test/parse.ts

```diff
@@ -0,0 +1,156 @@
+                });
+
+            // if the error is the start of a multiline error
+            } else if (errorLine instanceof MultilineErrorLine) {
+
+                // iterate through the MultilineErrorLines until we get to an EndErrorLine
+                for (let nextLineNo = lineNo + 1; ; ++nextLineNo) {
+                    if (!validErrorMarkupContinuation(errorLinesForCodeLines, nextLineNo)) {
+                        throw lintSyntaxError(
+                            `Error mark starting at ${errorStartPos.line}:${errorStartPos.col} does not end correctly.`
+                        );
+                    } else {
+                        const nextErrorLine = errorLinesForCodeLines[nextLineNo].shift();
+
+                        // if end of multiline error, add it it list of errors
+                        if (nextErrorLine instanceof EndErrorLine) {
+                            lintErrors.push({
+                                startPos: errorStartPos,
+                                endPos: { line: nextLineNo, col: nextErrorLine.endCol },
+                                message: messageSubstitutions[nextErrorLine.message] || nextErrorLine.message
+                            });
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    lintErrors.sort(errorComparator);
+
+    return lintErrors;
+}
+
+export function createMarkupFromErrors(code: string, lintErrors: LintError[]) {
+    lintErrors.sort(errorComparator);
+
+    const codeText = code.split("\n");
+    const errorLinesForCodeText: ErrorLine[][] = codeText.map(() => []);
+
+    for (const error of lintErrors) {
+        const {startPos, endPos, message} = error;
+
+        if (startPos.line === endPos.line) {  // single line error
+            errorLinesForCodeText[startPos.line].push(new EndErrorLine(
+                startPos.col,
+                endPos.col,
+                message
+            ));
+        } else {  // multiline error
+            errorLinesForCodeText[startPos.line].push(new MultilineErrorLine(startPos.col));
+            for (let lineNo = startPos.line + 1; lineNo < endPos.line; ++lineNo) {
+                errorLinesForCodeText[lineNo].push(new MultilineErrorLine(0));
+            }
+            errorLinesForCodeText[endPos.line].push(new EndErrorLine(0, endPos.col, message));
+        }
+    }
+
+    const finalText = combineCodeTextAndErrorLines(codeText, errorLinesForCodeText);
+    return finalText.join("\n");
+}
+/* tslint:enable:object-literal-sort-keys */
+
+function combineCodeTextAndErrorLines(codeText: string[], errorLinesForCodeText: ErrorLine[][]) {
+    return codeText.reduce((finalText, code, i) => {
+        finalText.push(code);
+        finalText.push(...errorLinesForCodeText[i].map((line) => lineObjToText(line, code)));
+        return finalText;
+    }, <string[]>[]);
+}
+
+function separateErrorLinesFromCodeLines(lines: Line[]) {
```

Comment:
how about `mapCodeLinesToErrorLines`? your sparse array below essentially works as a lookup table from each line of code to its list of errors